### DISCUSSION TITLE
ai.djl.engine.EngineException: No deep learning engine found

### DISCUSSION THREAD

**Discussion by drei34** [deepjavalibrary/djl#1956] [Q&A]
Hi,

I am trying to use the library through IntelliJ and I see that the standard way to do this is given by the information [here](https://docs.djl.ai/engines/pytorch/pytorch-engine/index.html). I am on maxOS. If I am allowed to visit external sites other than maven or some proximity repo I can use the below; it seems this goes out to the internet to fetch pytorch and therefore I cannot use this: 

```
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-engine</artifactId>
    <version>0.18.0</version>
    <scope>runtime</scope>
</dependency>
```

The workaround if you can't do this is (maybe I misunderstand) to use the below given that you know your architecture. However, this does not work and gives `ai.djl.engine.EngineException: No deep learning engine found` ... Is there something else I have to do? This solution is in the link provided.

```
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-native-cpu</artifactId>
    <classifier>osx-x86_64</classifier>
    <version>1.11.0</version>
    <scope>runtime</scope>
</dependency>
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-native-jni</artifactId>
    <version>1.11.0-0.18.0</version>
    <scope>runtime</scope>
</dependency>
```

My understanding is I should be able to use either the first suggestion or this second one and both should work, but the first will go and get pytorch while the second has it already and will not go and get it. Lastly, on maven I see "pytorch-jni" but no "pytorch-native-jni". Is there a type in the first link provided on your site? I.e., should the artifactId be "pytorch-jni"?

**Comment by frankfliu** [comment]
@drei34 

You need all those dependencies for offline use case. See this example: https://github.com/deepjavalibrary/djl-demo/tree/master/development/fatjar

**Reply by drei34** [reply]
@frankfliu By all these dependencies you mean the below? These alone do not seem to work. Sorry to be clear what I am assuming is that the below package pytorch - I have access to them. The issue with the one called ` <artifactId>pytorch-engine</artifactId>` is it wants to grab pytorch from another site and that's not allowed ... Thank you in advance!

```
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-native-cpu</artifactId>
    <classifier>osx-x86_64</classifier>
    <version>1.11.0</version>
    <scope>runtime</scope>
</dependency>
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-native-jni</artifactId>
    <version>1.11.0-0.18.0</version>
    <scope>runtime</scope>
</dependency>
```

**Comment by frankfliu** [comment]
@drei34 

You must include `<artifactId>pytorch-engine</artifactId>`. This package contains DJL's PyTorch engine adapter implementation.
If you have `pytorch-native-cpu-1.12.1-osx-x86_64.jar` in your classpath, DJL won't download from internet.

**Reply by drei34** [reply]
@frankfliu Thanks so much again. Just so I understand and I want to confirm for others in the future (correct me please if wrong). The dependencies should be as below and in the code before the model load there should be a line like `System.setProperty("ai.djl.pytorch.pytorch-engine", "pytorch-native-cpu-1.12.1-osx-x86_64.jar");` ... Also, note that the last dependency below has artifactId `pytorch-jni` and NOT `pytorch-native-jni`. Is this a typo? Is there an easy way to put this question on the webpage [link](https://docs.djl.ai/engines/pytorch/pytorch-engine/index.html#macos) ... Might be useful to people trying not to download or who can't download anything other than the dependencies below ... Or I guess they can search here ...

```
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-engine</artifactId>
    <version>0.18.0</version>
    <scope>runtime</scope>
</dependency>
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-native-cpu</artifactId>
    <classifier>osx-x86_64</classifier>
    <version>1.11.0</version>
    <scope>runtime</scope>
</dependency>
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-jni</artifactId>
    <version>1.11.0-0.18.0</version>
    <scope>runtime</scope>
</dependency>
```

**Comment by frankfliu** [comment]
I created a [PR](https://github.com/deepjavalibrary/djl/pull/1959) to fix the typo.
